File: app/src/test/java/fr/free/nrw/commons/PageTitleTest.java
package fr.free.nrw.commons;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.net.URLEncoder;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;

@RunWith(RobolectricTestRunner.class)
@Config(constants = BuildConfig.class, sdk = 21, application = TestCommonsApplication.class)
public class PageTitleTest {
    @Test
    public void displayTextShouldNotBeUnderscored() {
        assertThat(new PageTitle("Ex_1 ").getDisplayText(),
                is("Ex 1"));
    }

    @Test
    public void moreThanTwoColons() {
        assertThat(new PageTitle("File:sample:a.jpg").getPrefixedText(),
                is("File:Sample:a.jpg"));
    }

    @Test
    public void getTextShouldReturnWithoutNamespace() {
        assertThat(new PageTitle("File:sample.jpg").getText(),
                is("Sample.jpg"));
    }


    @Test
    public void capitalizeNameAfterNamespace() {
        assertThat(new PageTitle("File:sample.jpg").getPrefixedText(),
                is("File:Sample.jpg"));
    }

    @Test
    public void prefixedTextShouldBeUnderscored() {
        assertThat(new PageTitle("Ex 1 ").getPrefixedText(),
                is("Ex_1"));
    }

    @Test
    public void getMobileUriForTest() {
        assertThat(new PageTitle("Test").getMobileUri().toString(),
                is(BuildConfig.MOBILE_HOME_URL + "Test"));
    }

    @Test
    public void spaceBecomesUnderscoreInUri() {
        assertThat(new PageTitle("File:Ex 1.jpg").getCanonicalUri().toString(),
                is(BuildConfig.HOME_URL + "File:Ex_1.jpg"));
    }

    @Test
    public void leaveSubpageNamesUncapitalizedInUri() {
        assertThat(new PageTitle("User:Ex/subpage").getCanonicalUri().toString(),
                is(BuildConfig.HOME_URL + "User:Ex/subpage"));
    }

    @Test
    public void unicodeUri() throws Throwable {
        assertThat(new PageTitle("User:例").getCanonicalUri().toString(),
                is(BuildConfig.HOME_URL + "User:" + URLEncoder.encode("例", "utf-8")));
    }
}

File: app/src/test/kotlin/fr/free/nrw/commons/PageTitleTest.kt
package fr.free.nrw.commons

import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import java.net.URLEncoder

@RunWith(RobolectricTestRunner::class)
@Config(constants = BuildConfig::class, sdk = [21], application = TestCommonsApplication::class)
class PageTitleTest {
    @Test
    fun displayTextShouldNotBeUnderscored() {
        assertEquals("Ex 1", PageTitle("Ex_1").displayText)
        assertEquals("Ex 1", PageTitle("Ex_1 ").displayText)
        assertEquals("Ex 1", PageTitle("Ex 1").displayText)
        assertEquals("Ex 1", PageTitle("Ex 1 ").displayText)
        assertEquals("File:Ex 1 2", PageTitle("File:Ex_1_2 ").displayText)
    }

    @Test
    fun prefixedTextShouldBeUnderscored() {
        assertEquals("Ex_1", PageTitle("Ex_1").prefixedText)
        assertEquals("Ex_1", PageTitle("Ex_1 ").prefixedText)
        assertEquals("Ex_1", PageTitle("Ex 1").prefixedText)
        assertEquals("Ex_1", PageTitle("Ex 1 ").prefixedText)
        assertEquals("File:Ex_1_2", PageTitle("File:Ex 1 2 ").prefixedText)
    }

    @Test
    fun fileNameWithOneColon() {
        val pageTitle = PageTitle("File:sample:a.jpg")
        assertEquals("File:Sample:a.jpg", pageTitle.prefixedText)
        assertEquals("File:Sample:a.jpg", pageTitle.displayText)
        assertEquals("Sample:a.jpg", pageTitle.text)
    }

    @Test
    fun fileNameWithMoreThanOneColon() {
        var pageTitle = PageTitle("File:sample:a:b.jpg")
        assertEquals("File:Sample:a:b.jpg", pageTitle.prefixedText)
        assertEquals("File:Sample:a:b.jpg", pageTitle.displayText)
        assertEquals("Sample:a:b.jpg", pageTitle.text)

        pageTitle = PageTitle("File:sample:a:b:c.jpg")
        assertEquals("File:Sample:a:b:c.jpg", pageTitle.prefixedText)
        assertEquals("File:Sample:a:b:c.jpg", pageTitle.displayText)
        assertEquals("Sample:a:b:c.jpg", pageTitle.text)
    }

    @Test
    fun keyShouldNotIncludeNamespace() {
        val pageTitle = PageTitle("File:Sample.jpg")
        assertEquals("Sample.jpg", pageTitle.text)
    }

    @Test
    fun capitalizeNamespace() {
        val pageTitle = PageTitle("file:Sample.jpg")
        assertEquals("File:Sample.jpg", pageTitle.prefixedText)
        assertEquals("File:Sample.jpg", pageTitle.displayText)
    }

    @Test
    fun capitalizeKey() {
        val pageTitle = PageTitle("File:sample.jpg")
        assertEquals("File:Sample.jpg", pageTitle.prefixedText)
        assertEquals("File:Sample.jpg", pageTitle.displayText)
        assertEquals("Sample.jpg", pageTitle.text)
    }

    @Test
    fun getMobileUriForTest() {
        val pageTitle = PageTitle("Test")
        assertEquals(BuildConfig.MOBILE_HOME_URL + "Test", pageTitle.mobileUri.toString())
    }

    @Test
    fun spaceBecomesUnderscoreInUri() {
        val pageTitle = PageTitle("File:Ex 1.jpg")
        assertEquals(BuildConfig.HOME_URL + "File:Ex_1.jpg", pageTitle.canonicalUri.toString())
    }

    @Test
    fun leaveSubpageNamesUncapitalizedInUri() {
        val pageTitle = PageTitle("User:Ex/subpage")
        assertEquals(BuildConfig.HOME_URL + "User:Ex/subpage", pageTitle.canonicalUri.toString())
    }

    @Test
    fun unicodeUri() {
        val pageTitle = PageTitle("User:例")
        assertEquals(BuildConfig.HOME_URL + "User:" + URLEncoder.encode("例", "utf-8"), pageTitle.canonicalUri.toString())
    }
}
